packages/block-editor/src/components/responsive-block-control/index.js

```diff
@@ -0,0 +1,85 @@
+/**
+ * External dependencies
+ */
+import { uniqueId } from 'lodash';
+
+/**
+ * WordPress dependencies
+ */
+import { __, sprintf } from '@wordpress/i18n';
+
+import { Fragment, useState } from '@wordpress/element';
+
+import {
+	ToggleControl,
+} from '@wordpress/components';
+
+export function ResponsiveBlockControlLabel( { property, device } ) {
+	const accessibleLabel = sprintf( __( 'Controls the %s property for %s devices.' ), property, device );
+	const descId = uniqueId();
+	return (
+		<Fragment>
+			<span aria-describedby={ `rbc-desc-${ descId }` }>
+				{ device }
+			</span>
+			<span className="screen-reader-text" id={ `rbc-desc-${ descId }` }>{ accessibleLabel }</span>
+		</Fragment>
+	);
+}
+
+function ResponsiveBlockControl( props ) {
+	const { legend, property, toggleLabel, responsiveControlsActive = false, renderDefaultControl, defaultLabel = __( 'All' ), devices = [ __( 'Desktop' ), __( 'Tablet' ), __( 'Mobile' ) ], renderResponsiveControls } = props;
+
+	const [ isResponsiveMode, setIsResponsiveMode ] = useState( responsiveControlsActive );
```

Comment:
We keeping the responsive mode as a local state, but when the flag is toggled the parent needs to know about this state change, otherwise how is the change going to be persisted?

I think the parent should pass something like isResponsiveMode, and an onIsResponsiveMode change and the parent should be responsible for handling and storing changes.

Reply:
I've also been wrestling with this one.

You'll note until recently this was controlled by the consuming "parent" component. Then in [this commit, I moved it to become component/local state](https://github.com/WordPress/gutenberg/pull/16790/commits/1a0966b95d8469747640dc9579cae58634229873). 

This was because it felt odd that the consumer needed to provide a means to toggle the UI. I felt the state of the toggle should be internal (thereby avoiding the need for the user to have to know about this implementation detail) but we should expose a callback to notify when it changes and also a means to set the default state.

Note that currently, I've tried to find the best of both worlds. By default you don't need to handle the toggle manually when utilising the component - it is handled internally and "just works" (this is what I'd expect to happen as a user). However, you can optionally set the boolean `responsiveControlsActive` prop to determine the initial state of the UI (ie: open/closed responsive controls). This allows us to persist an open/closed state between renders and/or editor initialisations. I probably need a decent test case for this scenario.

It's simple enough to roll this back to my original implementation (which is what I think you are recommending) but I am concerned about making this component too hard to utilise.

Reply:
I've revised this a little. Let me know what you think.

Reply:
> This was because it felt odd that the consumer needed to provide a means to toggle the UI.

If the consumer passes isResponsiveMode and onIsResponsiveModeChange the consumer is not passing a means to toggle the UI it is passing if the current mode is responsive and a handler to change it. It is not even aware that the mode is changed by using a toggle.
If responsiveControlsActive changes, the component will not reflect that change, this seems problematic is it makes it impossible to apply a change in the flag from a third party plugin if the component is rendered.

Reply:
To be clear

`responsiveControlsActive` - this sets the initial state of the toggle. This allows the consuming component to set the initial mode to responsive or not.

`onIsResponsiveModeChange` - this is a callback purely to "notify" the consuming component that the responsive mode has changed to the value provided as an argument (ie: on/off). It doesn't do any _setting_.

In addition, there is also the _internal_ state value `isResponsiveMode`. This is used to manage the state of the responsive toggle internally. The consuming component has no means to alter this value except on the initial render where the initial value of the state is set by `responsiveControlsActive` prop.

My thinking was that it was too much to ask the consuming component to manage the internal toggle state of the UI. I thought that by managing internally it would make the component easier to use. I recognised that the consuming component would need to be able to set the initial toggle state which is why I went for the `responsiveControlsActive` prop.

That said, recent experience has led me to believe that almost all components that involve user input need to be fully controlled components. I think this is what [you were driving at](https://github.com/WordPress/gutenberg/pull/16790#discussion_r331606578) and I'm coming around to that idea, even if it makes the component more labourious to consume.